src/Log/Engine/BaseLog.php

```diff
@@ -61,6 +68,23 @@ public function __construct(array $config = [])
         if (!empty($this->_config['types']) && empty($this->_config['levels'])) {
             $this->_config['levels'] = (array)$this->_config['types'];
         }
+
+        // Default to DefaultFormatter for BC with custom loggers that don't set `formatter` in $_defaultConfig
+        $formatterConfig = $this->_config['formatter'] ?? DefaultFormatter::class;
+        if (!is_array($formatterConfig)) {
+            $formatterConfig = ['className' => $formatterConfig];
+        }
+
+        /** @var class-string<\Cake\Log\Formatter\AbstractFormatter> $formatterClass */
+        $formatterClass = $formatterConfig['className'];
+        $this->formatter = new $formatterClass($formatterConfig);
+        if (!$this->formatter instanceof AbstractFormatter) {
```

Comment:
You could use `ReflectionClass` to check this before constructing it. That would also allow you to skip the type annotation as you'd be able to use `$reflection->newInstance()`

Reply:
Seems it made no difference to psalm